File: .github/scripts/test-parent-owned-vendoring.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "fileutils"
require "open3"
require "rbconfig"
require "tmpdir"

SOURCE_ROOT = File.expand_path("../..", __dir__)
FIXTURE = File.join(SOURCE_ROOT, ".github/fixtures/profiles/script-assisted")
VALIDATOR = File.join(SOURCE_ROOT, ".github/scripts/validate-skill-repository.rb")
SKILL_FILES = %w[SKILL.md scripts/normalize.rb].freeze
FAILURES = []

def capture(*command, chdir:, env: {})
  Open3.capture3(env, *command, chdir: chdir)
end

def run!(*command, chdir:, env: {})
  stdout, stderr, status = capture(*command, chdir: chdir, env: env)
  return stdout if status.success?

  raise "command failed: #{command.inspect}; status=#{status.exitstatus.inspect}; " \
        "stdout=#{stdout.inspect}; stderr=#{stderr.inspect}"
end

def validate(target, outside)
  capture(
    RbConfig.ruby,
    VALIDATOR,
    target,
    chdir: outside,
    env: { "RUBYOPT" => nil }
  )
end

def expect_validation_success(target, outside)
  stdout, stderr, status = validate(target, outside)
  return if status.success? && stderr.empty? &&
            stdout.include?("Agent Skill repository structure and profile contracts are valid.")

  FAILURES << "expected parent-owned vendored skill validation success; " \
              "status=#{status.exitstatus.inspect}, stdout=#{stdout.inspect}, stderr=#{stderr.inspect}"
end

def expect_gitlink_rejection(target, outside)
  _stdout, stderr, status = validate(target, outside)
  diagnostic = "Operational resource gitlinks are not allowed: scripts/index-only-link"
  if status.success?
    FAILURES << "expected parent-index gitlink rejection"
  elsif !stderr.include?(diagnostic)
    FAILURES << "expected diagnostic #{diagnostic.inspect}; stderr=#{stderr.inspect}"
  end
end

def git_index_bytes(repository)
  index_path = run!("git", "rev-parse", "--git-path", "index", chdir: repository).strip
  index_path = File.expand_path(index_path, repository) unless index_path.start_with?(File::SEPARATOR)
  File.binread(index_path)
end

def tracked_skill_entries(repository, prefix)
  output = run!("git", "ls-files", "--stage", "--", prefix, chdir: repository)
  output.lines.to_h do |line|
    match = line.match(/\A(\d+)\s+[0-9a-f]+\s+\d+\t(.+)\n?\z/)
    raise "unexpected git ls-files record: #{line.inspect}" unless match

    [match[2], match[1]]
  end
end

def content_map(root)
  SKILL_FILES.to_h { |relative| [relative, File.binread(File.join(root, relative))] }
end

def exercise_helper(target)
  input_path = File.join(target, "input.txt")
  output_path = File.join(target, "output.txt")
  File.binwrite(input_path, "alpha  \r\nbeta\t\r\n")
  before = File.binread(input_path)

  stdout, stderr, status = capture(
    RbConfig.ruby,
    "scripts/normalize.rb",
    "input.txt",
    "output.txt",
    chdir: target
  )
  output = File.binread(output_path) if File.file?(output_path)
  return if status.success? && stdout == "output.txt\n" && stderr.empty? &&
            output == "alpha\nbeta\n" && File.binread(input_path) == before

  FAILURES << "helper execution failed; status=#{status.exitstatus.inspect}, " \
              "stdout=#{stdout.inspect}, stderr=#{stderr.inspect}, output=#{output.inspect}"
end

Dir.mktmpdir("parent-owned-vendoring") do |workspace|
  outside = File.join(workspace, "outside")
  source = File.join(workspace, "concrete-skill-source")
  archive = File.join(workspace, "concrete-skill.tar")
  extracted = File.join(workspace, "archive-extracted")
  parent = File.join(workspace, "parent-project")
  skill_prefix = ".agents/skills/line-normalization-helper"
  target = File.join(parent, skill_prefix)

  FileUtils.mkdir_p([outside, source, extracted, parent])
  FileUtils.cp_r("#{FIXTURE}/.", source, preserve: true)

  run!("git", "init", "--quiet", chdir: source)
  run!("git", "config", "user.name", "Parent-owned Vendoring Fixture", chdir: source)
  run!("git", "config", "user.email", "fixture@example.invalid", chdir: source)
  run!("git", "add", ".", chdir: source)
  run!("git", "commit", "--quiet", "-m", "Create concrete skill", chdir: source)
  run!(
    "git", "archive", "--format=tar", "--prefix=line-normalization-helper/",
    "--output=#{archive}", "HEAD",
    chdir: source
  )
  run!("tar", "-xf", archive, "-C", extracted, chdir: workspace)

  run!("git", "init", "--quiet", chdir: parent)
  run!("git", "config", "user.name", "Parent Project Fixture", chdir: parent)
  run!("git", "config", "user.email", "parent@example.invalid", chdir: parent)
  File.write(File.join(parent, "README.md"), "# Parent project\n", encoding: "UTF-8")
  run!("git", "add", "README.md", chdir: parent)
  run!("git", "commit", "--quiet", "-m", "Create parent project", chdir: parent)

  FileUtils.mkdir_p(target)
  FileUtils.cp_r(
    "#{File.join(extracted, 'line-normalization-helper')}/.",
    target,
    preserve: true
  )
  run!("git", "add", "--", skill_prefix, chdir: parent)
  run!("git", "commit", "--quiet", "-m", "Vendor concrete skill", chdir: parent)

  expected_entries = SKILL_FILES.to_h do |relative|
    [File.join(skill_prefix, relative), "100644"]
  end
  actual_entries = tracked_skill_entries(parent, skill_prefix)
  FAILURES << "parent index does not own the expected regular files: #{actual_entries.inspect}" \
    unless actual_entries == expected_entries

  FAILURES << "vendored bytes differ from the committed source" \
    unless content_map(target) == content_map(source)
  FAILURES << "vendored target unexpectedly contains Git metadata" \
    if File.exist?(File.join(target, ".git"))
  FAILURES << "vendored target retained an archive wrapper" \
    if Dir.exist?(File.join(target, "line-normalization-helper"))

  discovered_root = run!("git", "rev-parse", "--show-toplevel", chdir: target).strip
  FAILURES << "validator target did not discover the parent worktree" \
    unless File.realpath(discovered_root) == File.realpath(parent)

  index_before = git_index_bytes(parent)
  expect_validation_success(target, outside)
  index_after = git_index_bytes(parent)
  FAILURES << "successful validation modified the parent index" unless index_after == index_before
  exercise_helper(target)

  parent_head = run!("git", "rev-parse", "HEAD", chdir: parent).strip
  gitlink_path = File.join(skill_prefix, "scripts", "index-only-link")
  run!(
    "git", "update-index", "--add", "--cacheinfo",
    "160000,#{parent_head},#{gitlink_path}",
    chdir: parent
  )
  FAILURES << "index-only gitlink unexpectedly exists on the filesystem" \
    if File.exist?(File.join(parent, gitlink_path))

  gitlink_index_before = git_index_bytes(parent)
  expect_gitlink_rejection(target, outside)
  gitlink_index_after = git_index_bytes(parent)
  FAILURES << "failed validation modified the parent index" \
    unless gitlink_index_after == gitlink_index_before
end

unless FAILURES.empty?
  FAILURES.each { |failure| warn failure }
  exit 1
end

puts "Parent-owned vendoring smoke tests passed."

File: .github/scripts/test_parent_owned_vendoring.py
#!/usr/bin/env python3
"""Prove validation of a concrete Skill vendored into a parent Git worktree."""

from __future__ import annotations

import os
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

SOURCE_ROOT = Path(__file__).resolve().parents[2]
FIXTURE = SOURCE_ROOT / ".github/fixtures/profiles/script-assisted"
VALIDATOR = SOURCE_ROOT / "template/.github/scripts/validate_skill_repository.py"
SKILL_FILES = ["SKILL.md", "scripts/normalize.py"]
FAILURES: list[str] = []


def clean_env(extra: dict[str, str] | None = None) -> dict[str, str]:
    env = os.environ.copy()
    for key in ("PYTHONPATH", "GIT_DIR", "GIT_INDEX_FILE", "GIT_WORK_TREE"):
        env.pop(key, None)
    if extra:
        env.update(extra)
    return env


def capture(
    *command: str, cwd: Path, extra_env: dict[str, str] | None = None
) -> subprocess.CompletedProcess[str]:
    return subprocess.run(
        list(command),
        cwd=cwd,
        env=clean_env(extra_env),
        text=True,
        encoding="utf-8",
        capture_output=True,
        check=False,
    )


def run_or_raise(*command: str, cwd: Path) -> str:
    result = capture(*command, cwd=cwd)
    if result.returncode != 0:
        raise RuntimeError(
            f"command failed: {command!r}; status={result.returncode!r}; "
            f"stdout={result.stdout!r}; stderr={result.stderr!r}"
        )
    return result.stdout


def validate(target: Path, outside: Path) -> subprocess.CompletedProcess[str]:
    return capture(
        sys.executable,
        str(VALIDATOR),
        str(target),
        cwd=outside,
    )


def git_index_bytes(repository: Path) -> bytes:
    index_text = run_or_raise(
        "git", "rev-parse", "--git-path", "index", cwd=repository
    ).strip()
    index_path = Path(index_text)
    if not index_path.is_absolute():
        index_path = repository / index_path
    return index_path.read_bytes()


def tracked_skill_entries(repository: Path, prefix: str) -> dict[str, str]:
    output = run_or_raise(
        "git", "ls-files", "--stage", "--", prefix, cwd=repository
    )
    entries: dict[str, str] = {}
    for line in output.splitlines():
        metadata, path = line.split("\t", 1)
        mode, _sha, _stage = metadata.split(" ", 2)
        entries[path] = mode
    return entries


def content_map(root: Path) -> dict[str, bytes]:
    return {relative: (root / relative).read_bytes() for relative in SKILL_FILES}


def exercise_helper(target: Path) -> None:
    input_path = target / "input.txt"
    output_path = target / "output.txt"
    input_path.write_bytes(b"alpha  \r\nbeta\t\r\n")
    before = input_path.read_bytes()
    result = capture(
        sys.executable,
        "scripts/normalize.py",
        "input.txt",
        "output.txt",
        cwd=target,
    )
    output = output_path.read_bytes() if output_path.is_file() else None
    if not (
        result.returncode == 0
        and result.stdout == "output.txt\n"
        and result.stderr == ""
        and output == b"alpha\nbeta\n"
        and input_path.read_bytes() == before
    ):
        FAILURES.append(
            "helper execution failed; "
            f"status={result.returncode!r}, stdout={result.stdout!r}, "
            f"stderr={result.stderr!r}, output={output!r}"
        )


def main() -> int:
    with tempfile.TemporaryDirectory(prefix="parent-owned-vendoring") as temporary:
        workspace = Path(temporary)
        outside = workspace / "outside"
        source = workspace / "concrete-skill-source"
        archive = workspace / "concrete-skill.tar"
        extracted = workspace / "archive-extracted"
        parent = workspace / "parent-project"
        skill_prefix = ".agents/skills/line-normalization-helper"
        target = parent / skill_prefix

        for path in (outside, source, extracted, parent):
            path.mkdir(parents=True, exist_ok=True)
        shutil.copytree(FIXTURE, source, dirs_exist_ok=True, symlinks=True)

        run_or_raise("git", "init", "--quiet", cwd=source)
        run_or_raise(
            "git", "config", "user.name", "Parent-owned Vendoring Fixture", cwd=source
        )
        run_or_raise(
            "git", "config", "user.email", "fixture@example.invalid", cwd=source
        )
        run_or_raise("git", "add", ".", cwd=source)
        run_or_raise(
            "git", "commit", "--quiet", "-m", "Create concrete skill", cwd=source
        )
        run_or_raise(
            "git",
            "archive",
            "--format=tar",
            "--prefix=line-normalization-helper/",
            f"--output={archive}",
            "HEAD",
            cwd=source,
        )
        run_or_raise("tar", "-xf", str(archive), "-C", str(extracted), cwd=workspace)

        run_or_raise("git", "init", "--quiet", cwd=parent)
        run_or_raise(
            "git", "config", "user.name", "Parent Project Fixture", cwd=parent
        )
        run_or_raise(
            "git", "config", "user.email", "parent@example.invalid", cwd=parent
        )
        (parent / "README.md").write_text("# Parent project\n", encoding="utf-8")
        run_or_raise("git", "add", "README.md", cwd=parent)
        run_or_raise(
            "git", "commit", "--quiet", "-m", "Create parent project", cwd=parent
        )

        target.mkdir(parents=True)
        shutil.copytree(
            extracted / "line-normalization-helper",
            target,
            dirs_exist_ok=True,
            symlinks=True,
        )
        run_or_raise("git", "add", "--", skill_prefix, cwd=parent)
        run_or_raise(
            "git", "commit", "--quiet", "-m", "Vendor concrete skill", cwd=parent
        )

        expected_entries = {
            f"{skill_prefix}/{relative}": "100644" for relative in SKILL_FILES
        }
        actual_entries = tracked_skill_entries(parent, skill_prefix)
        if actual_entries != expected_entries:
            FAILURES.append(
                "parent index does not own the expected regular files: "
                f"{actual_entries!r}"
            )
        if content_map(target) != content_map(source):
            FAILURES.append("vendored bytes differ from the committed source")
        if (target / ".git").exists():
            FAILURES.append("vendored target unexpectedly contains Git metadata")
        if (target / "line-normalization-helper").is_dir():
            FAILURES.append("vendored target retained an archive wrapper")

        discovered_root = Path(
            run_or_raise("git", "rev-parse", "--show-toplevel", cwd=target).strip()
        )
        if discovered_root.resolve() != parent.resolve():
            FAILURES.append("validator target did not discover the parent worktree")

        index_before = git_index_bytes(parent)
        result = validate(target, outside)
        if not (
            result.returncode == 0
            and result.stderr == ""
            and "Agent Skill repository structure and profile contracts are valid."
            in result.stdout
        ):
            FAILURES.append(
                "expected parent-owned vendored skill validation success; "
                f"status={result.returncode!r}, stdout={result.stdout!r}, "
                f"stderr={result.stderr!r}"
            )
        if git_index_bytes(parent) != index_before:
            FAILURES.append("successful validation modified the parent index")
        exercise_helper(target)

        parent_head = run_or_raise("git", "rev-parse", "HEAD", cwd=parent).strip()
        gitlink_path = f"{skill_prefix}/scripts/index-only-link"
        run_or_raise(
            "git",
            "update-index",
            "--add",
            "--cacheinfo",
            f"160000,{parent_head},{gitlink_path}",
            cwd=parent,
        )
        if (parent / gitlink_path).exists():
            FAILURES.append("index-only gitlink unexpectedly exists on the filesystem")

        gitlink_index_before = git_index_bytes(parent)
        result = validate(target, outside)
        diagnostic = "Operational resource gitlinks are not allowed: scripts/index-only-link"
        if result.returncode == 0:
            FAILURES.append("expected parent-index gitlink rejection")
        elif diagnostic not in result.stderr:
            FAILURES.append(
                f"expected diagnostic {diagnostic!r}; stderr={result.stderr!r}"
            )
        if git_index_bytes(parent) != gitlink_index_before:
            FAILURES.append("failed validation modified the parent index")

    if FAILURES:
        for failure in FAILURES:
            print(failure, file=sys.stderr)
        return 1
    print("Parent-owned vendoring smoke tests passed.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
